[BEAM-498] Replace ParDo with MapElements and FlatMapElements where possible

kennknowles opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

The commits ended up having fairly separate topics, but can be reviewed individually or as a medium-sized change.

1. The first commit replaces `ParDo` with `MapElements` and `FlatMapElements` where it is easy to do so.
2. While debugging, I noticed that `DoFn` used a less-powerful form of `TypeDescriptor` and switched trivially to the enhanced version.
3. The root cause of issues with `MapElements` and `FlatMapElements` was a lack of use of the input type descriptor. Making it available involved a moderate refactor. In the process I broke some tests to do with display data and fixed them plus enhancements to display data for `SimpleFunction`.

If reviewers insist, I can try to alter this commit history.

R: @bjchambers AND @swegner

swegner commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java:
```diff
@@ -85,42 +85,57 @@ private MissingOutputTypeDescriptor(SerializableFunction<InputT, OutputT> fn) {
-  private final SerializableFunction<InputT, OutputT> fn;
-  private final transient TypeDescriptor<OutputT> outputType;
+  private final SimpleFunction<InputT, OutputT> fn;
+  private final Class<?> fnClass; // for display data purposes
```
Careful, storing `Class<?>` instances can be problematic. Anonymous classes backed by Java 8 lambdas will explode during serialization. See [CombineJava8Test#testLambdaSerialization()](https://github.com/apache/incubator-beam/blob/master/sdks/java/java8tests/src/test/java/org/apache/beam/sdk/transforms/CombineJava8Test.java#L144). In Combine.java, we [store the `DisplayData.Item<?>` instead](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java#L127).

Recommend writing a Java8 test for this class as well. I wonder: would be it be easy to write some FindBugs-like static analysis that would catch `Serializable` classes with non-transient `Class<?>` instance fields?

kennknowles replied:
Thank you! I did not identify the cause of the failure, but this was certainly the issue. Whatever I store does have to be serialized and non-transient, because of the unfortunate issue with `ParDo`. Your idea solved it. I'm going to drop the commit from this PR, please review #757 that is focused on just that.

swegner commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java:
```diff
@@ -148,7 +148,7 @@ public Integer apply(Integer input) {
 
   @Test
   public void testSimpleFunctionDisplayData() {
```
Add validation for registered `SimpleFunction<,>` display data:

```java
@Override
public void populateDipslayData(DisplayData.Builder builder) {
  builder.add(DispayData.item("foo", "bar"));
}

// ...

DisplayData displayData = DisplayData.from(simpleMap);
assertThat(displayData, hasDisplayItem("mapFn", simpleFn.getClass()));
assertThat(displayData, hasDisplayItem("foo", "bar));
```

kennknowles replied:
Added these tests, but moved the commit to a different branch that is not yet in a PR. Once #757 is merged, it will unblock this PR as well as enable opening a PR to add display data to `SimpleFunction`, which is no longer on the critical path for me.

swegner commented:
This is small enough to review as two commits, but I'd recommend fixing up the history on merge to separate out the changes in two commits.

kennknowles commented:
R: -@swegner love the feedback but this PR may or may not remain interesting to you

kennknowles commented:
PTAL. Rebased and fixed up.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java:
```diff
@@ -25,7 +25,7 @@
  */
 public class MapElements<InputT, OutputT>
-extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+extends PTransform<PCollection<? extends InputT>, PCollection<OutputT>> {
```
Were the changes to MapElements not in a different PR already?

kennknowles replied:
This particular change was not. There were cases that could not be ported without it, since the new generic makes it the same type as `ParDo`.

kennknowles commented:
This actually looks like a Jenkins race condition or some such. The errors in that build are not part of this PR. Rebasing to kick...

kennknowles commented:
R: -@bjchambers I had added you since you were very in-the-loop on both the new `DoFn` and display data, but now this PR has nothing to do with those and I've sent more than enough your way.

R: @aljoscha care to take a look? The tests have been passing for a while, but with some Travis timeouts. I just rebased to force Travis to run again in its new, faster, configuration.

bjchambers commented:
LGTM
